url647b55da7c_Nikhilkumar13_emuSr_tgz-pJ8-MainJ8/src/frontend/handlers.java

```diff
@@ -0,0 +1,959 @@
+      }
+
+    void paste()
+      {
+        Transferable cliptran = FrontEnd.clip.getContents(this);
+        try
+        {
+            String sel = (String) cliptran.getTransferData(DataFlavor.stringFlavor);
+            FrontEnd.activepane.replaceSelection(sel);
+        } catch (Exception ex)
+        {
+            Logger.getLogger(handlers.class.getName()).log(Level.SEVERE, null, ex);
+        }
+
+      }
+
+    void run()
+      {
+        FrontEnd.exceptionraised = 0;
+        new Memory();
+        backend.Register.resetRegisters();
+        clean_branchtable();
+        clean_memtable();
+         FrontEnd.activepane.getHighlighter().removeAllHighlights();
+       
+
+        backend.Register.r[15].b = backend.ScanFile.endOfProgram;
+
+//        sActionPerformed(evt);
+        BufferedWriter out = null;
+        File fin = null;
+
+        int id = FrontEnd.EditorPane.getSelectedIndex();
+        if (FrontEnd.EditorPane.getToolTipTextAt(id) != null)
+        {
+            //FrontEnd.statuswindow.append("Saving File...\n");
+            FrontEnd.appendToPane(FrontEnd.statuswindow,"Saving File\n",Color.BLACK);
+            //frontend.FrontEnd.statuswindow.setCaretPosition(frontend.FrontEnd.statuswindow.getText().length());
+            try
+            {
+                FrontEnd.filepath = FrontEnd.EditorPane.getToolTipTextAt(id);
+                fin = new File(FrontEnd.filepath);
+                out = new BufferedWriter(new FileWriter(fin));
+                out.write(FrontEnd.activepane.getText());
+            } catch (IOException ex)
+            {
+                //FrontEnd.statuswindow.append("Error occured while saving file...\n");
+                FrontEnd.appendToPane(FrontEnd.statuswindow,"Error Occured While Saving File...\n",Color.BLACK);
+//                FrontEnd.statuswindow.append(ex.getStackTrace().toString());
+                FrontEnd.appendToPane(FrontEnd.statuswindow,ex.getStackTrace().toString(),Color.BLACK);
+                //frontend.FrontEnd.statuswindow.setCaretPosition(frontend.FrontEnd.statuswindow.getText().length());
+            } finally
+            {
+                try
+                {
+                    out.close();
+                } catch (IOException ex)
+                {
+//                    FrontEnd.statuswindow.append("Error occured while saving file...\n");
+//                    FrontEnd.statuswindow.append(ex.getStackTrace().toString());
+                    FrontEnd.appendToPane(FrontEnd.statuswindow,"Error Occured While Saving File...\n",Color.BLACK);
+//                FrontEnd.statuswindow.append(ex.getStackTrace().toString());
+                FrontEnd.appendToPane(FrontEnd.statuswindow,ex.getStackTrace().toString(),Color.BLACK);
+                    
+                    //frontend.FrontEnd.statuswindow.setCaretPosition(frontend.FrontEnd.statuswindow.getText().length());
+                    frontend.FrontEnd.exceptionraised++;
+                }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 293:306, 518:529.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.